### ISSUE TITLE
[Feature Request] Infer `NDBuffer` Rank From DimList

### ISSUE THREAD

**Issue by lsh** [modular/modular#4]
In most APIs, the rank of a multidimensional array would be assumed to be equivalent to the shape.

I noticed that the invocation for `NDBuffer` is `NDBuffer[rank, shape, dtype]()`. It would be ergonomic to have a constructor that assumes rank from shape, so it could just be `NDBuffer[DimList(2,3), DType.ui32]()`,
or even better `NDBuffer[(2,3), DType.ui32]()`

**Comment by Mogball** [comment]
Thanks for the issue!

This is a current limitation of how we handle dependent parameters, and there are other types and functions that suffer from this. The key issue is that the `shape` parameter depends on `rank`, and parameters are resolved left-to-right, so `rank` has to be defined first. That said, I think this is something we should support, since it is natural and intuitive to infer `rank` from `shape`.

**Comment by lattner** [comment]
@abduld this seems like we can just directly improve NDBuffer.  Could we change it from:

```
struct NDBuffer[
    rank: Int,
    shape: DimList,
    type: DType,
]:
```

to (pseudo code):

```
struct NDBuffer[
    shape: DimList,
    type: DType,
]:
  alias rank = len(shape) 
  ...
```

DimList doesn't currently have a `__len__` method, but its underlying storage type does.

**Comment by abduld** [comment]
Yes, we can certainly improve (and will improve) the API for NDBuffer. Thanks @lsh for the suggestion

**Comment by josiahls** [comment]
@lsh @abduld Looking for simple git issues to solve. Is this ever going to be worked on? I looked at today's implementation of ndbuffer:
```mojo
struct NDBuffer[
    mut: Bool, //,
    dtype: DType,
    rank: Int,
    origin: Origin[mut],
    shape: DimList = DimList.create_unknown[rank](),
    strides: DimList = DimList.create_unknown[rank](),
    *,
    alignment: Int = 1,
    address_space: AddressSpace = AddressSpace.GENERIC,
    exclusive: Bool = True,
]
```
I looks like it was decided to do the reverse and use rank to determine the shape / strides.

**Comment by lsh** [comment]
NDBuffer is being removed in favor of LayoutTensor, which doesn't have this problem. This issue can be closed.
